backend/mttnet/site.go

```diff
@@ -0,0 +1,888 @@
+	info, ok := rpcpeer.FromContext(ctx)
+	if !ok {
+		return cid.Cid{}, fmt.Errorf("BUG: no peer info in context for grpc")
+	}
+
+	pid, err := peer.Decode(info.Addr.String())
+	if err != nil {
+		return cid.Cid{}, err
+	}
+
+	return peer.ToCid(pid), nil
+}
+
+func randStr(length int) string {
+	randomBytes := make([]byte, 32)
+	_, err := rand.Read(randomBytes)
+	if err != nil {
+		panic(err)
+	}
+	return base32.StdEncoding.EncodeToString(randomBytes)[:length]
+}
+
+// Client dials a remote peer if necessary and returns the RPC client handle.
+func (srv *Server) Client(ctx context.Context, device cid.Cid) (site.WebSiteClient, error) {
+	n, ok := srv.Node.Get()
+	if !ok {
+		return nil, fmt.Errorf("Node not ready yet")
+	}
+	pid, err := peer.FromCid(device)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := n.Connect(ctx, n.p2p.Peerstore().PeerInfo(pid)); err != nil {
+		return nil, err
+	}
+	return n.client.DialSite(ctx, pid)
+}
+
+func (srv *Server) checkPermissions(ctx context.Context, requiredRole site.Member_Role, params ...interface{}) (cid.Cid, bool, interface{}, error) {
+	n, ok := srv.Node.Get()
+	if !ok {
+		return cid.Cid{}, false, nil, fmt.Errorf("Node not ready yet")
+	}
+
+	remoteHostname, err := getRemoteSiteFromHeader(ctx)
+	if err != nil && srv.hostname == "" { // no headers and not a site
+		return cid.Cid{}, false, nil, fmt.Errorf("This node is not a site, please provide a proper headers to proxy the call to a remote site: %w", err)
+	}
+
+	acc := n.me.AccountID()
+	n.log.Debug("Check permissions", zap.String("Site hostname", srv.hostname), zap.String("remoteHostname", remoteHostname), zap.Error(err))
+	if err == nil && srv.hostname != remoteHostname && srv.hostname != "" {
+		return acc, false, nil, fmt.Errorf("Hostnames don't match. This site's hostname is [%s] but called with headers [%s]", srv.hostname, remoteHostname)
+	}
+
+	if err == nil && srv.hostname != remoteHostname && srv.hostname == "" { //This call is intended to be proxied so its site's duty to check permission
+		// proxy to remote
+		if len(params) == 0 {
+			n.log.Error("Headers found, meaning this call should be proxied, but remote function params not provided")
+			return acc, false, nil, fmt.Errorf("In order to proxy a call (headers found) you need to provide a valid proxy func and a params")
+		}
+		n.log.Debug("Headers found, meaning this call should be proxied and authentication will take place at the remote site", zap.String(string(MttHeader), remoteHostname))
+
+		pc, _, _, _ := runtime.Caller(1)
+		proxyFcnList := strings.Split(runtime.FuncForPC(pc).Name(), ".")
+		proxyFcn := proxyFcnList[len(proxyFcnList)-1]
+		res, errInterface := srv.proxyToSite(ctx, remoteHostname, proxyFcn, params...)
+		if errInterface != nil {
+			err, ok := errInterface.(error)
+			if !ok {
+				return acc, true, res, fmt.Errorf("Proxied call returned unknown second parameter. Error type expected")
+			}
+			return acc, true, res, err
+		}
+		return acc, true, res, nil
```

Comment:
@juligasa, I'm a bit confused about what's happening here. It seems like proxying the requests may not be the responsibility of the `mttnet` package. From what I can see, it looks like you may have combined the proxying part and serving part inside the same handler with the help of the reflect magic. I noticed that the site server is exposed inside the daemon grpc instance and over libp2p as well.

I think having two separate implementations for the Site interface would be more appropriate and explicit, in fact I thought this was what you were initially going for. Have you found any issues with this method?

In my opinion, it would be much more clear to have one implementation of the Site service that is exposed over libp2p and handles all the logic, and a separate implementation of the same interface that only does proxying. This way, all the header logic would be handled in the proxying implementation, and no reflect magic would be needed.

Reply:
But then I would have the same logic (the proxying logic) replicated in all the functions, right?. With reflections the proxy logic is in one place, no matter how many new functions I want to add in the future.

Reply:
@juligasa this is true, and I think it's a great example where the Don't Repeat Yourself principle is not worth it. These functions would be so simple and easily copy-pasteatble that having reflection magic and dynamic invocations is not giving much benefit, IMO:

1. Check the target hostname in the header.
2. Create an RPC client for the hostname.
3. Invoke the corresponding method on the client.
4. Return the result.

I think the rule of thumb in Go is to think twice whenever you reach for reflection, especially for dynamically invoking functions. The most common and most legitimate use of reflection is probably parsing data structures. Overall, in Go there's a fair bit of repetition in many places :) Even in this code, all of these handlers look very similar: `checkPermissions()`, then if request is proxied do something, else do something else; and it's totally fine, IMO.

Having to proxy each function explicitly also gives a bit more control, e.g. we don't want to proxy the `RedeemInviteToken` call because it's handled by the `AddSite` call behind the scenes, so we can return an error when this call is attempted to proxy.

If you feel I'm talking bullshit (in the end it's probably more matter of taste, although avoiding function calls with reflect is common in the Go community), please, at least write some good comment explaining what's happening there and what is the reasoning.

Reply:
I see. However, the `checkPermissions()` call is inevitable since not all calls require the same permissions. Also, I think the use of reflections here is safe since the function name to call is taken from the very same codebase where the replection is executed, so no real panic at runtime may happen. Even if the function is not implemented remotely, it will return an error instead of panicking.